stats: add unit support to histogram

troshko111 opened the pull request:
Signed-off-by: Taras Roshko <troshko@netflix.com>

Description: Add unit to the histogram type so that the stat sinks can disambiguate among various types of histograms (some Metric stores support timer vs non-timer histograms and generally speaking there already are histograms measuring bytes and microseconds, but it's all assumed to be millis on flush
Risk Level: Low
Testing: Unit tests
Docs Changes: N/A
Release Notes: Included
Fixes: #8164

troshko111 commented on include/envoy/stats/histogram.h:
```diff
@@ -67,33 +68,68 @@ class HistogramStatistics {
+  /**
+   * @return the unit symbol.
+   */
+  virtual std::string unit_symbol() const {
```
`unit_symbol` -> `unitSymbol`.

troshko111 commented on test/mocks/stats/mocks.h:
```diff
@@ -229,14 +234,19 @@ class MockParentHistogram : public MockMetric<ParentHistogram> {
 
   MOCK_CONST_METHOD0(used, bool());
   MOCK_METHOD1(recordValue, void(uint64_t value));
+  MOCK_CONST_METHOD0(unit, Histogram::Unit());
```
reorder (before `recordValue`).

troshko111 commented on test/mocks/stats/mocks.h:
```diff
@@ -205,13 +205,18 @@ class MockHistogram : public MockMetric<Histogram> {
   ~MockHistogram() override;
 
   MOCK_METHOD1(recordValue, void(uint64_t value));
+  MOCK_CONST_METHOD0(unit, Histogram::Unit());
```
Reorder (before recordValue).

troshko111 commented on test/mocks/stats/mocks.cc:
```diff
@@ -45,6 +46,7 @@ MockParentHistogram::MockParentHistogram() {
       store_->deliverHistogramToSinks(*this, value);
     }
   }));
+  ON_CALL(*this, unit()).WillByDefault(ReturnPointee(&unit_));
```
Reorder (before `recordValue`).

troshko111 commented on test/mocks/stats/mocks.cc:
```diff
@@ -36,6 +36,7 @@ MockHistogram::MockHistogram() {
       store_->deliverHistogramToSinks(*this, value);
     }
   }));
+  ON_CALL(*this, unit()).WillByDefault(ReturnPointee(&unit_));
```
Reorder (before `recordValue`).

troshko111 commented on include/envoy/stats/timespan.h:
```diff
@@ -24,35 +24,57 @@ class CompletableTimespan {
 public:
-  TimespanWithUnit(Histogram& histogram, TimeSource& time_source)
-      : time_source_(time_source), histogram_(histogram), start_(time_source.monotonicTime()) {}
+  HistogramCompletableTimespan(Histogram& histogram, TimeSource& time_source)
```
If the histogram is rejected by stats matcher, a null histogram is returned with an `Unspecified` unit so it is not safe to fail like this.

troshko111 commented:
@jmarantz 
Release builds are failing due to stat size checking tests:

`StatsThreadLocalStoreTestNoFixture.MemoryWithoutTlsFakeSymbolTable` and `StatsThreadLocalStoreTestNoFixture.MemoryWithTlsFakeSymbolTable` are 16 bytes off

`ClusterMemoryTestRunner.MemoryLargeClusterSizeWithFakeSymbolTable` and 
`ClusterMemoryTestRunner.MemoryLargeClusterSizeWithRealSymbolTable` are 32 bytes off

I don't know exactly how these are computed but a small size difference is expected, let me know if I can adjust the asserts.

Other things I'd like your opinion on:

1) `HistogramCompletableTimespan` is kind of ugly with these asserts, one thing I was thinking of is to introduce a method on histogram like `quantity()` returning an enum with values `Unspecified, Information, Time` and assert on `quantity == Time` instead of the switch in the ctor.
2) Multiple null histograms cached in fields in `ThreadLocalStoreImpl` vs enum to null histogram map (this requires either a copy ctor on the null histogram class or a move ctor which seemed like a bad trade off so I stored them in fields in this PR)
3) I'd like to update the metrics service to send the unit along with the bucket data, this is a forward/backward compatible change and should not break anyone.

jmarantz commented:
Yes if you are adding some stat data of value, like an enum to a histogram, that seems totally worth it and you should just bump the limit.

You can also test locally with 'blaze test -c opt --test_env=ENVOY_MEMORY_TEST_EXACT=true' so you can get the test working locally before pushing to CI.

I will start reviewing your PR; sorry for the delay.

jmarantz commented on include/envoy/event/dispatcher.h:
```diff
@@ -30,8 +30,8 @@ namespace Event {
 #define ALL_DISPATCHER_STATS(HISTOGRAM)                                                            \
-  HISTOGRAM(loop_duration_us)                                                                      \
-  HISTOGRAM(poll_delay_us)
+  HISTOGRAM(loop_duration, Microseconds)                                                           \
```
I see how this is a positive change. However I wonder if this might be a breaking change for monitoring setups for some deployments. How will we message this out and manage this transition?

troshko111 replied:
I've been thinking about this change before making it and came to a conclusion that however nasty this may be, the alternative is no better (potentially worse).

This is a breaking change as people definitely have alerts set up for these. If left as-is, `statsd` change is not possible (automatically append the actually correct symbol) and Metric sinks supporting generic histograms vs timers will report milli-milliseconds, micro-milliseconds and other nonsense (they would take values in seconds and scale using metric prefixes and display the scaled value so the metric name suffix will mismatch). It's also simply put messy. Some of the reported histograms right now are rather confusing already (Redis command latency is reported in millis or micros depending on a config value) so alerting on these is error-prone at best.

Having said that we may create a table of old -> new metric names for `statsd` and for `non-statsd` sinks so people can do a low effort find & replace instead of figuring it out? This can be published with release itself as a pre-read.

Thoughts?

jmarantz replied:
@htuch @alyssawilk should this be considered an API change and versioned as such?

alyssawilk replied:
To sanity check, this is changing the variable names for all statsd variables?  I definitely think that's something that folks should be able to migrate to carefully and not have to fix as they pull HEAD.
Ideally it's the sort of thing we could "runtime" guard and do away with but I don't think it's the sort of thing you can flip live.  Given that I'd lean towards a config option which we deprecate to (eventually) avoid code cruft

troshko111 replied:
@alyssawilk this affects histogram names, I've compiled the changes to make this easier to understand:


### Histogram name changes for non-statsd sinks (unit symbol suffix dropped if present)

| Group       |                        Old | New                    |
| :---------- | -------------------------: | :--------------------- |
| Cluster     |    `upstream_cx_length_ms` | `upstream_cx_length`   |
| Cluster     |   `upstream_cx_connect_ms` | `upstream_cx_connect`  |
| Dispatcher  |            `poll_delay_us` | `poll_delay`           |
| Dispatcher  |         `loop_duration_us` | `loop_duration`        |
| Dubbo       |          `request_time_ms` | `request_time`         |
| Dynamo      |         `upstream_rq_time` | same                   |
| Dynamo      |   `upstream_rq_time_{i}xx` | same                   |
| Dynamo      | `upstream_rq_time_unknown` | same                   |
| HTTPConnMan |       `downstream_rq_time` | same                   |
| HTTPConnMan |  `downstream_cx_length_ms` | `downstream_cx_length` |
| Listener    |  `downstream_cx_length_ms` | `downstream_cx_length` |
| Mongo       |                  `time_ms` | `time`                 |
| Mongo       |               `reply_size` | same                   |
| Mongo       |            `reply_time_ms` | `reply_time`           |
| Redis       |                  `latency` | same                   |
| Redis       |         `upstream_rq_time` | same                   |
| Server      |   `initialization_time_ms` | `initialization_time`  |
| Thrift      |          `request_time_ms` | `request_time`         |
| UserAgent   |  `downstream_cx_length_ms` | `downstream_cx_length` |
| ZK          |         `<opcode>_latency` | same                   |
| ZK          | `connect_response_latency` | same                   |

### Histogram name changes for statsd sinks (unit symbol suffix appended where absent)

| Group       |                        Old | New                                         |
| :---------- | -------------------------: | :------------------------------------------ |
| Cluster     |    `upstream_cx_length_ms` | same                                        |
| Cluster     |   `upstream_cx_connect_ms` | same                                        |
| Dispatcher  |            `poll_delay_us` | same                                        |
| Dispatcher  |         `loop_duration_us` | same                                        |
| Dubbo       |          `request_time_ms` | same                                        |
| Dynamo      |         `upstream_rq_time` | `upstream_rq_time_ms`                       |
| Dynamo      |   `upstream_rq_time_{i}xx` | `upstream_rq_time_{i}xx_ms`                 |
| Dynamo      | `upstream_rq_time_unknown` | `upstream_rq_time_unknown_ms`               |
| HTTPConnMan |       `downstream_rq_time` | `downstream_rq_time_ms`                     |
| HTTPConnMan |  `downstream_cx_length_ms` | same                                        |
| Listener    |  `downstream_cx_length_ms` | same                                        |
| Mongo       |                  `time_ms` | same                                        |
| Mongo       |               `reply_size` | `reply_size_b`                              |
| Mongo       |            `reply_time_ms` | same                                        |
| Redis       |                  `latency` | `latency_us` or `latency_ms` (configurable) |
| Redis       |         `upstream_rq_time` | `upstream_rq_time_ms`                       |
| Server      |   `initialization_time_ms` | same                                        |
| Thrift      |          `request_time_ms` | same                                        |
| UserAgent   |  `downstream_cx_length_ms` | same                                        |
| ZK          |         `<opcode>_latency` | `<opcode>_latency_ms`                       |
| ZK          | `connect_response_latency` | `connect_response_latency_ms`               |

mattklein123 replied:
IMO it's a non-starter to change the stat names. This is too disruptive to existing users.

I haven't yet reviewed this PR but can we please not change any stat names?

troshko111 replied:
Sure, I can revert the name changes and the `statsd` suffixing so all histogram names are always reported as-is. Is this going to be the recommendation for new histograms as well?

Right now Envoy own histograms are all suffixed (except 1 `downstream_rq_time` in `http_conn_man`) but extensions are all over the place. If you decide to always take the name as-is, we can maybe add a formatter check for name suffixes not matching the provided unit?

mattklein123 replied:
Sure, I think it's great to require good name hygiene for new stats, I just don't think it's worth the end-user pain to change the names of the existing ones (though I agree it would be nice).

troshko111 replied:
ok so to recap:
1) Keep all existing names as-is.
2) Do not autosuffix in `statsd`.
3) Have a `check_format` rule ensuring no new histograms are suffixed but whitelist all existing.

I will address the remaining comments and send 1 commit to undo the renames so it's easier to review.

troshko111 replied:
I will leave the `statsd` to the sink maintainers, but recommendation is that since `statsd` assumes all histograms are millis, all other time values should be scaled to millis on flush unless they are suffixed. This is the minimally invasive fix to reinstall sanity. This way things like Redis `latency` would be scaled to millis and things like `poll_delay_us` would be reported as-is but it would be clear on the UI it's micros (due to the name). And for all new stats they can be assumed to be millis, no suffixed, scaled to millis if necessary.

htuch replied:
A little late to the discussion, but I agree that this is too disruptive. However, stats are not (today) part of the Envoy API, so they don't live by its versioning guidelines. Is there a way we can add versioning to the stats interface to allow it to evolve over time? Should we provide a canonical proto definition that will allow it to move alongside the API major versions, or perhaps on its own clock?

troshko111 replied:
I overall agree this is a breaking change and probably can't be released just like this but *I'd really welcome any strategy for dealing with metric name migrations* as little things here and there keep being added over time and at some point consistency and clarity is lost in the name of backward compatibility and at that point it's hard to say what causes more bugs - vague metrics and wrong alerts set up based on misinterpretation or broken alerts due to the name changes. It's not a pressing problem but would be could if there was a solution at some point.

jmarantz commented on include/envoy/stats/histogram.h:
```diff
@@ -67,33 +68,70 @@ class HistogramStatistics {
+  /**
+   * @return the unit symbol.
+   */
+  virtual std::string unitSymbol() const {
```
s/std::string/absl::string_view/ to save having to allocate and copy string data on every call.

troshko111 replied:
Fixed.

jmarantz commented on include/envoy/stats/histogram.h:
```diff
@@ -67,33 +68,70 @@ class HistogramStatistics {
+  virtual Unit unit() const PURE;
+
+  /**
+   * @return the unit symbol.
```
I don't think this method is overridden anywhere, so there's no need to make it virtual.

However it doesn't quite look like it belongs as a concrete function in an interface.

Maybe it could be moved to a utility library like source/common/stats/utility.h

troshko111 replied:
Fixed.

jmarantz commented on include/envoy/stats/histogram.h:
```diff
@@ -67,33 +68,70 @@ class HistogramStatistics {
+      return "ms";
+    }
+
+    return "";
```
return "unknown" and/or assert?

troshko111 replied:
Fixed.

jmarantz commented on include/envoy/stats/timespan.h:
```diff
@@ -24,35 +24,61 @@ class CompletableTimespan {
+    switch (histogram.unit()) {
+    case Histogram::Unit::Unspecified:
+    case Histogram::Unit::Bytes:
+      ASSERT(0);
```
I like the change to templatize only the method. Can we then move rest of the impls into an _impl.cc file?

troshko111 replied:
Fixed.

jmarantz commented on source/common/http/user_agent.cc:
```diff
@@ -21,7 +21,8 @@ void UserAgent::completeConnectionLength(Stats::Timespan& span) {
   }
 
-  scope_->histogram(prefix_ + "downstream_cx_length_ms").recordValue(span.getRawDuration().count());
+  scope_->histogram(prefix_ + "downstream_cx_length", Stats::Histogram::Unit::Milliseconds)
```
can you do me a favor and put:

// TODO(jmarantz): use stat-names here.

troshko111 replied:
Done.

jmarantz commented on source/common/stats/histogram_impl.h:
```diff
@@ -81,12 +81,15 @@ class HistogramImpl : public HistogramImplHelper {
   SymbolTable& symbolTable() override { return parent_.symbolTable(); }
 
 private:
+  Unit unit_;
```
FWIW this is likely where the memory-consumption difference comes from, and this makes sense. No response needed on this; you can just adjust the expected value as needed.

troshko111 replied:
Yep, the mem alignment changed and it was off both ways actually. Adjusted.

jmarantz commented on source/common/stats/isolated_store_impl.h:
```diff
@@ -77,6 +90,7 @@ template <class Base> class IsolatedStatsCache {
   StatNameHashMap<RefcountPtr<Base>> stats_;
   Allocator alloc_;
   AllocatorImportMode alloc_import_;
+  AllocatorUnit alloc_unit_;
```
I think with this we've given up all pretense of sharing functors. Maybe just rename them all:

CounterAllocator
GaugeAllocator
HistogramAllocator

troshko111 replied:
Fixed.

jmarantz commented on source/common/stats/thread_local_store.cc:
```diff
@@ -25,7 +25,10 @@ ThreadLocalStoreImpl::ThreadLocalStoreImpl(Allocator& alloc)
       stats_matcher_(std::make_unique<StatsMatcherImpl>()), heap_allocator_(alloc.symbolTable()),
       null_counter_(alloc.symbolTable()), null_gauge_(alloc.symbolTable()),
-      null_histogram_(alloc.symbolTable()) {}
+      null_histogram_(alloc.symbolTable(), Histogram::Unit::Unspecified),
```
do we need to have Null Histograms with the different units? Maybe we can just have null histograms always use unspecified units?

troshko111 replied:
We actually do, some histograms are used in `Timespans` and if you were to create a rejection rule to reject some of them, a `NullHistogram` of type `Unspecified` would get used in the `Timespan` and crash the process (ASSERT on the histogram type).

I wanted to reduce the ugliness by creating a `enum` -> `NullHistogram` map but realized the type is neither copyable nor movable so ended up just adding a couple of fields.

Technically speaking we can have only two - one `NullHistogram` for time and one for the rest but somehow it felt hacky (and no compiler warning on enum changes)?

troshko111 replied:
Fixed the way you suggested, pushing in a bit.

jmarantz commented on source/common/stats/histogram_impl.h:
```diff
@@ -97,17 +100,19 @@ class HistogramImpl : public HistogramImplHelper {
   bool used() const override { return false; }
   SymbolTable& symbolTable() override { return symbol_table_; }
 
+  Unit unit() const override { return unit_; };
```
i'm thinking just hard-code returning Unspecified here?

troshko111 replied:
See the reply on `thread_local_store.cc`.

troshko111 replied:
Fixed, pushing in a bit.

jmarantz commented on source/extensions/filters/network/dubbo_proxy/stats.h:
```diff
@@ -33,7 +33,7 @@ namespace DubboProxy {
   COUNTER(response_success)                                                                        \
   GAUGE(request_active, Accumulate)                                                                \
-  HISTOGRAM(request_time_ms)
+  HISTOGRAM(request_time, Milliseconds)
```
again we need to be careful to message out how we avoid breaking monitoring for people.

troshko111 replied:
<Consolidated the rename discussion in the OP>

jmarantz commented on source/extensions/filters/network/redis_proxy/command_splitter_impl.h:
```diff
@@ -45,18 +45,18 @@ class Utility {
   COUNTER(total)                                                                                   \
   COUNTER(success)                                                                                 \
-  COUNTER(error)                                                                                   \
-  HISTOGRAM(latency)
```
what happened to this?

troshko111 replied:
This extension relied on runtime decision for microsecond vs millisecond precision of this histogram and since it's no longer possible to change the unit at run time I moved the declaration out of the macro to the decision point, see `InstanceImpl::addHandler` in `source/extensions/filters/network/redis_proxy/command_splitter_impl.cc`.

jmarantz commented on source/extensions/filters/network/thrift_proxy/stats.h:
```diff
@@ -29,7 +29,7 @@ namespace ThriftProxy {
   COUNTER(response_success)                                                                        \
   GAUGE(request_active, Accumulate)                                                                \
-  HISTOGRAM(request_time_ms)
+  HISTOGRAM(request_time, Milliseconds)
```
ditto for messaging out the update.

troshko111 replied:
Consolidating the rename discussion in the OP.

jmarantz commented on source/extensions/stat_sinks/common/statsd/statsd.cc:
```diff
@@ -67,8 +67,12 @@ void UdpStatsdSink::flush(Stats::MetricSnapshot& snapshot) {
-  // For statsd histograms are all timers.
-  const std::string message(fmt::format("{}.{}:{}|ms{}", prefix_, getName(histogram),
+  // For statsd histograms are all timers, append the ASCII SI symbol to disambiguate.
+  std::string si_name = histogram.unitSymbol().length() > 0
```
taste-test: WDYT of writing this as:
```
std::string si_name(getName(histogram));
if (histogram.unit() != Histogram::Unit::Undefined) {
  absl::StrAppend(&si_name, "_", histogramUnitSymbol();
}
```

This seems easier to read than the ternary, and avoids calling unitSymbol() twice, which (in current PR) returns a string, causing an alloca and a copy.

troshko111 replied:
Fixed, no strong opinion on the readability (slightly like ternaries better here) but definitely agree on the double invocation.

jmarantz commented on source/extensions/stat_sinks/common/statsd/statsd.cc:
```diff
@@ -67,8 +67,12 @@ void UdpStatsdSink::flush(Stats::MetricSnapshot& snapshot) {
+                            ? fmt::format("{}_{}", getName(histogram), histogram.unitSymbol())
+                            : getName(histogram);
+
+  const std::string message(fmt::format("{}.{}:{}|ms{}", prefix_, si_name,
```
why do we have hardcoded "ms" here *and* use unitSymbol() above?

Should we just ASSERT that the units is milliseconds here?

Or add comments to explain the strategy here.

troshko111 replied:
statsd uses "ms" to specify the metric is a histogram, under the hood it does not have histogram vs timer distinction and everything is a timer. Added comments to disambiguate.

jmarantz commented on source/extensions/stat_sinks/common/statsd/statsd.h:
```diff
@@ -88,9 +88,12 @@ class TcpStatsdSink : public Stats::Sink {
-                                                 std::chrono::milliseconds(value));
+    // For statsd histograms are all timers, append the ASCII SI symbol to disambiguate.
+    std::string si_name = histogram.unitSymbol().length() > 0
+                              ? fmt::format("{}_{}", histogram.name(), histogram.unitSymbol())
```
i'm seeing this pattern in a few places. factor it out to source/common/stats/utility.h.

troshko111 replied:
Fixed.

jmarantz commented on source/server/server.h:
```diff
@@ -63,7 +63,7 @@ namespace Server {
   GAUGE(uptime, Accumulate)                                                                        \
   GAUGE(version, NeverImport)                                                                      \
-  HISTOGRAM(initialization_time_ms)
+  HISTOGRAM(initialization_time, Milliseconds)
```
again we need a story for how users migrate their monitoring when this gets merged.

troshko111 replied:
Consolidating the rename discussion in the OP.

jmarantz commented on test/common/stats/stat_test_utility.cc:
```diff
@@ -83,12 +83,12 @@ void forEachSampleStat(int num_clusters, std::function<void(absl::string_view)>
 
   // These are the other stats that appear in the admin /stats request when made
   // prior to any requests.
-  static const char* other_stats[] = {"http.admin.downstream_cx_length_ms",
```
Consider adding an exclusion for creating histograms ending with "_ms" in check_format.py, so that none of these sneak in later.

troshko111 replied:
Great idea, added.

jmarantz commented on test/mocks/stats/mocks.h:
```diff
@@ -204,14 +204,19 @@ class MockHistogram : public MockMetric<Histogram> {
   bool decRefCount() override { return refcount_helper_.decRefCount(); }
   uint32_t use_count() const override { return refcount_helper_.use_count(); }
 
+  Unit unit_;
```
````
unit_{Unit::Unspecified};
```
then you don't have to add those 'unspecified' inits all over the test code.

troshko111 replied:
I've checked the test changes and almost all of the instantiations go through store/scope. There's only a handful of the ones using mocks directly and some of them set to `non-Unspecified` anyway, to be consistent I'd rather set them to `Unspecified` explicitly as well. Again it's some usages in `test/extensions/stats_sinks/common/statsd/udp_statsd_test.cc` and `test/server/http/admin_test.cc` and 1 in server test. WDYT?

jmarantz replied:
I still think you should default-init the member-variable, either in the ctor or as an initializer.

troshko111 replied:
Ah fair point, missed that.

troshko111 commented:
Thanks a bunch for taking the time to review this @jmarantz!

I think I addressed the comments (minus the release messaging and some questions I had).

Let me know what your thoughts are on the questions 1-3, I think we should at least update the metric service contract to send the unit.

jmarantz commented on include/envoy/stats/timespan.h:
```diff
@@ -24,37 +24,31 @@ class CompletableTimespan {
   }
 
-private:
+protected:
```
leave this private: and move histogram_ into the impl. Actually I'd probably move time_source_ and start_ into the impl as well, otherwise this is not really an interface, but maybe it's not worth it, especially as this was the existing condition.

troshko111 replied:
Moved everything and added a virtual `elapsedMs` method to the interface (since I could not use templates), does this look better?

jmarantz commented:
Oh sorry I missed the 3 questions:

> HistogramCompletableTimespan is kind of ugly with these asserts, one thing I was thinking of is to introduce a method on histogram like quantity() returning an enum with values Unspecified, Information, Time and assert on quantity == Time instead of the switch in the ctor.

I didn't notice asserts in there. Maybe I'm missing something?

> Multiple null histograms cached in fields in ThreadLocalStoreImpl vs enum to null histogram map (this requires either a copy ctor on the null histogram class or a move ctor which seemed like a bad trade off so I stored them in fields in this PR)

I think what you have is OK, but I wanted to take one more shot at simplifying the null histograms. Can you just change the asserts to allow for the case of a null histogram? Say, by having an isNull() pure method in the Histogram interface that you can check?

> I'd like to update the metrics service to send the unit along with the bucket data, this is a forward/backward compatible change and should not break anyone.

ok sounds fine to me then.


My main blocker is I'd like to hear from @envoyproxy/senior-maintainers about the breaking changes to monitoring and how to message that and/or phase it.

jmarantz commented:
Leaving the names as-is sounds good to me. You could whitelist any existing names that are in violation; there's a few similar patterns in check_format.py.

troshko111 commented on source/common/stats/timespan_impl.cc:
```diff
@@ -0,0 +1,47 @@
+    return;
+  }
+
+  switch (histogram.unit()) {
```
@jmarantz I was talking about this switch, histogram can define another enum say `Quantity` and we can add a util method to derive `quantity` from `unit` so code would only need to do
`ASSERT(Util::quantity(histogram.unit()) == Histogram::Quantity::Time)`

or even add it to the histogram interface itself.

worth it or leave as-is?

jmarantz replied:
You could also just do
```
  ASSERT(histogram.unit() == Histogram::Unit::Microseconds || histogram.unit() == Histogram::Unit::Milliseconds);
```
that is (a) less lines of code than what you have and (b) will print out exactly what the problem is in the assertion method if it fails. If there was some other reason to create Util::quantity() you could use that, but I wouldn't create it for this.

troshko111 replied:
Thanks for your take on this, I made it a switch since a modern compiler checks if all cases were tested so if a new time unit is added, the compilation would fail and it would be obvious what places need to be updated, if I just check for micros and millis here and one day someone adds nanos, this will introduce a silent bug. No strong preference either way, just let me know if you still prefer ` ASSERT(histogram.unit() == Histogram::Unit::Microseconds || histogram.unit() == Histogram::Unit::Milliseconds);` compared to a full switch and I'll update it..

jmarantz replied:
Great point. I agree.

troshko111 commented:
> I didn't notice asserts in there. Maybe I'm missing something?

> Oh sorry I missed the 3 questions:
> 
> > HistogramCompletableTimespan is kind of ugly with these asserts, one thing I was thinking of is to introduce a method on histogram like quantity() returning an enum with values Unspecified, Information, Time and assert on quantity == Time instead of the switch in the ctor.
> 
> I didn't notice asserts in there. Maybe I'm missing something?
> 
> > Multiple null histograms cached in fields in ThreadLocalStoreImpl vs enum to null histogram map (this requires either a copy ctor on the null histogram class or a move ctor which seemed like a bad trade off so I stored them in fields in this PR)
> 
> I think what you have is OK, but I wanted to take one more shot at simplifying the null histograms. Can you just change the asserts to allow for the case of a null histogram? Say, by having an isNull() pure method in the Histogram interface that you can check?
> 
> > I'd like to update the metrics service to send the unit along with the bucket data, this is a forward/backward compatible change and should not break anyone.
> 
> ok sounds fine to me then.
> 
> My main blocker is I'd like to hear from @envoyproxy/senior-maintainers about the breaking changes to monitoring and how to message that and/or phase it.

1) Tagged you directly in the diff to explain what I was thinking in more detail.
2) Simplified the null histograms as per your suggestion (I was thinking about similar approach but was not sure you'd be fine with a virtual marker, I made it more generally useful so it does not look like a hack, let me know if that's what you meant)
3) Will push a separate commit to plumb through the metric service

Finally as per Matt's suggestion I'm reverting the name changes altogether and will leave a TODO in `statsd` for the maintainers to consider scaling unsuffixed histograms reported not in millis as these are a usability disaster for the users (`statsd` thinks it's millis, users think it's millis, the name has no indication it's not, but it's not). This will ensure backward-compat. Will tune the `check_format` to give everything existing a pass and bark about any future suffixes.

troshko111 commented:
`macOS` test failure seems intermittent (a timeout), other than that I think I've addressed everything. Will not be changing metric service as this is uses Prometheus proto contract so out of scope.

jmarantz commented on include/envoy/stats/histogram.h:
```diff
@@ -46,16 +46,17 @@ class HistogramStatistics {
-   * guaranteed to be the same length as supportedBuckets().
+   * Returns computed bucket values during the period. The vector contains an
+   * approximation of samples below each quantile bucket defined in
+   * supportedBuckets(). This vector is guaranteed to be the same length as
```
can you revert this change? It looks like it is a reformatting; might as well leave it as it was.

Do you have some kind of automatic comment reformatter in your IDE or something?

troshko111 replied:
`clangd` reformatting based on `.clang-format`, will revert.

troshko111 replied:
Seems like a `clang-format` glitch or something, reverted and saved and it did not reformat again but kept as-is.

jmarantz commented on include/envoy/stats/histogram.h:
```diff
@@ -46,16 +46,17 @@ class HistogramStatistics {
-   * of the number of samples in the histogram, it is not guaranteed that this will be
-   * 100% the number of samples observed.
+   * Returns number of values during the period. This number may be an
+   * approximation of the number of samples in the histogram, it is not
```
ditto.

jmarantz commented on include/envoy/stats/histogram.h:
```diff
@@ -67,33 +68,61 @@ class HistogramStatistics {
   /**
-   * This method is called during the main stats flush process for each of the histograms and used
-   * to merge the histogram values.
+   * This method is called during the main stats flush process for each of the
```
revert?

jmarantz commented on include/envoy/stats/histogram.h:
```diff
@@ -67,33 +68,61 @@ class HistogramStatistics {
 
 /**
- * A histogram that is stored in main thread and provides summary view of the histogram.
+ * A histogram that is stored in main thread and provides summary view of the
```
revert?

jmarantz commented on source/extensions/filters/network/redis_proxy/command_splitter_impl.h:
```diff
@@ -45,18 +46,18 @@ class Utility {
 struct CommandStats {
-  ALL_COMMAND_STATS(GENERATE_COUNTER_STRUCT, GENERATE_HISTOGRAM_STRUCT)
+  ALL_COMMAND_STATS(GENERATE_COUNTER_STRUCT)
+  Envoy::Stats::Histogram& latency_;
```
can you just make sure we have test-cases covering millis and micros for this?

troshko111 replied:
There are, it's in `test/extensions/filters/network/redis_proxy/command_splitter_impl_test.cc`,  all test check the millis and `RedisSingleServerRequestWithLatencyMicrosTest` specifically check the `micros`.

jmarantz commented on source/extensions/stat_sinks/common/statsd/statsd.cc:
```diff
@@ -67,7 +67,12 @@ void UdpStatsdSink::flush(Stats::MetricSnapshot& snapshot) {
 
 void UdpStatsdSink::onHistogramComplete(const Stats::Histogram& histogram, uint64_t value) {
-  // For statsd histograms are all timers.
+  // TODO: for statsd histograms are all timers in milliseconds, Envoy histograms are however not
```
TODO(username) or TODO(#issue)

jmarantz commented on source/extensions/stat_sinks/common/statsd/statsd.cc:
```diff
@@ -206,6 +211,7 @@ void TcpStatsdSink::TlsSink::onTimespanComplete(const std::string& name,
                                                 std::chrono::milliseconds ms) {
   // Ultimately it would be nice to perf optimize this path also, but it's not very frequent. It's
   // also currently not possible that this interleaves with any counter/gauge flushing.
+  // TODO: see the comment at UdpStatsdSink::onHistogramComplete.
```
ditto

jmarantz commented on test/extensions/stats_sinks/common/statsd/statsd_test.cc:
```diff
@@ -95,13 +95,50 @@ TEST_F(TcpStatsdSinkTest, BasicFlow) {
 
   NiceMock<Stats::MockHistogram> timer;
   timer.name_ = "test_timer";
+  timer.unit_ = Stats::Histogram::Unit::Unspecified;
```
you can drop this, right?

jmarantz commented on test/extensions/stats_sinks/common/statsd/statsd_test.cc:
```diff
@@ -95,13 +95,50 @@ TEST_F(TcpStatsdSinkTest, BasicFlow) {
+
+  NiceMock<Stats::MockHistogram> items;
+  items.name_ = "items";
+  items.unit_ = Stats::Histogram::Unit::Unspecified;
```
ditto

troshko111 replied:
Dropping it from everywhere but `SiSuffix` tests (3 of these total) since this is the whole point of the test to check whether the unit affects the suffix in the expected way and better to have it explicit in the test setup.

mattklein123 commented on include/envoy/stats/timespan.h:
```diff
@@ -18,43 +15,18 @@ class CompletableTimespan {
+   * Time elapsed since the creation of the timespan, in milliseconds.
    */
-  void complete() override { histogram_.recordValue(getRawDuration().count()); }
+  virtual uint64_t elapsedMs() PURE;
```
Can this return a `std::chrono::millseconds` and just be called `elapsed()`?

troshko111 replied:
Good point, fixed.

mattklein123 commented on source/common/stats/timespan_impl.cc:
```diff
@@ -0,0 +1,47 @@
+  switch (histogram.unit()) {
+  case Histogram::Unit::Unspecified:
+  case Histogram::Unit::Bytes:
+    ASSERT(0);
```
What is the purpose of this ASSERT? I would remove.

Edit: Sorry what is purpose of this entire switch statement?

jmarantz replied:
I think it's just to verify that this histogram has a time-based unit.  A comment to that effect would help, or maybe a helper method so it's self-documenting.

troshko111 replied:
Extracted into a method.

Correct, this is to ensure the histogram passed to the timespan is measuring time, if that's not the case, it's hard to say what `complete()` should record, how would we convert duration to say `bytes`? This is always a programming error but there's no elegant way to make this a compile time check AFAICT.

mattklein123 commented on source/common/stats/timespan_impl.cc:
```diff
@@ -0,0 +1,47 @@
+    return;
+  }
+
+  ASSERT(0);
```
this should be NOT_REACHED if this is required for compilation.

troshko111 replied:
My bad, did not notice this macro even though it was right there next to assert. Fixed throughout.

mattklein123 commented on source/common/stats/timespan_impl.cc:
```diff
@@ -0,0 +1,47 @@
+  switch (histogram_.unit()) {
+  case Histogram::Unit::Unspecified:
+  case Histogram::Unit::Bytes:
+    // Unreachable, asserted to measure time on construction.
```
Then use NOT_REACHED

mattklein123 commented on source/extensions/stat_sinks/common/statsd/statsd.cc:
```diff
@@ -67,7 +67,12 @@ void UdpStatsdSink::flush(Stats::MetricSnapshot& snapshot) {
 
 void UdpStatsdSink::onHistogramComplete(const Stats::Histogram& histogram, uint64_t value) {
-  // For statsd histograms are all timers.
+  // TODO(#8594): for statsd histograms are all timers in milliseconds, Envoy histograms are however
```
Personally, I would downgrade from TODO and just leave a note here in case this comes up again. Tons of people use Envoy with statsd and I have never heard anyone mention this problem with our stats histogram implementation. If/when someone actually cares about this we can revisit. Same below.

troshko111 replied:
Removed the TODO and left as a note. There are not many of these right now and they're all in extensions so I suspect the users would not complain but it's something to consider going forward.

mattklein123 commented on source/common/stats/timespan_impl.cc:
```diff
@@ -0,0 +1,47 @@
+HistogramCompletableTimespanImpl::HistogramCompletableTimespanImpl(Histogram& histogram,
+                                                                   TimeSource& time_source)
+    : time_source_(time_source), histogram_(histogram), start_(time_source.monotonicTime()) {
+  if (!histogram.active()) {
```
What is the purpose of the `active()` method in general and why do we check it here? This constructor doesn't do anything AFAICT so I'm confused about the purpose.

jmarantz replied:
This was likely the result of suggestion I made earlier to differentiate null histograms for the purpose of avoiding the asserts below.

But another thing that occurred to me is just add Histogram::Unit::NullHistogram to the enum and put that in the switch.

troshko111 replied:
I followed @jmarantz's suggestion here and made it an enum value instead, let me know what you all think, can revert back.

mattklein123 commented on source/common/stats/timespan_impl.cc:
```diff
@@ -0,0 +1,49 @@
+    return;
+  case Histogram::Unit::Unspecified:
+  case Histogram::Unit::Bytes:
+    ASSERT(0);
```
The intention here is to demonstrate that this is a programming error, right? 

If so, can you make this `ASSERT(false, "some text that is useful...")` so that in dev people will see the mistake they have made? Also, technically I think this should be a RELEASE_ASSERT since we will crash later on anyway with NOT_REACHED, right?

troshko111 replied:
Right, I've added an error message which hopefully would be useful and agree it's better as a release assert because if such error slips through the tests and debug builds it's better to fail with a clear message than panic. Fixed both.

mattklein123 commented:
Sorry looks like it needs a master merge.

/wait

mattklein123 commented:
@troshko111 friendly request to please never force push, just merge master and push commits. Thank you!

troshko111 commented:
> @troshko111 friendly request to please never force push, just merge master and push commits. Thank you!

Ah my bad! I thought merge conflicts were ok to rebase, I now see that GitHub makes it painful to review when the history changes (does not store snapshots to compare to), sorry for that.

qqustc commented:
@troshko111 
hi, I'm trying to add new envoy histograms in this PR https://github.com/envoyproxy/nighthawk/pull/378, e.g. `HISTOGRAM(latency_on_success_req_us, Microseconds)`

I got [check_format error](https://github.com/envoyproxy/envoy/blob/dcf34972d1bc15324835c40dfd7a780e8fc69d72/tools/code_format/check_format.py#L677-L682) because of the unit in the definition 
*"Don't suffix histogram names with the unit symbol, it's already part of the histogram object and unit-supporting sinks can use this information natively, other sinks can add the suffix automatically on flush should they prefer to do so."*

However build failed after I removed the unit from the definition `HISTOGRAM(latency_on_success_req_us)`
because [GENERATE_HISTOGRAM_STRUCT(NAME, UNIT)](https://github.com/envoyproxy/envoy/blob/dcf34972d1bc15324835c40dfd7a780e8fc69d72/include/envoy/stats/stats_macros.h#L41) requires unit in the histogram definition

So I'm confused, is the [check_format rule](https://github.com/envoyproxy/envoy/blob/dcf34972d1bc15324835c40dfd7a780e8fc69d72/tools/code_format/check_format.py#L677-L682) outdated? If not, what is the recommended way to define histogram? 

Thanks!
